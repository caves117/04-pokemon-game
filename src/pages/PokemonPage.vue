<template>
  <h1 v-if="!pokemon"> Wait</h1>
  <div v-else>
    <h1>¿Quién es este pokemón?</h1>
    <PokemonPicture 
    :pokemonId="pokemon.id" 
    :showPokemon="showPokemon"/>
    <PokemonOptions 
    :pokemons="pokemonArr" 
    @selection="checkAnswer"/>
    <template v-if="showAnswer" class="fade-in">
      <h2>{{message}}</h2>
      <button @click="newGame">Nuevo juego</button>
    </template>
  </div>

</template>

<script>

import PokemonPicture from '@/components/PokemonPicture.vue'
import PokemonOptions from '@/components/PokemonOptions.vue'

import {getPokemonOptions} from '@/helpers/getPokemonOptions'

export default {
    components: {
      PokemonPicture,
      PokemonOptions
    },
    data() {
      return {
        pokemonArr: [],
        pokemon: null,
        showPokemon: false,
        showAnswer: false,
        message: ''
      }
    },
    methods: {
      async mixPokemonArr() {
        this.pokemonArr = await getPokemonOptions()
        const rndInt = Math.floor(Math.random() * 4)
        this.pokemon = this.pokemonArr[rndInt]
      },
      checkAnswer(pokemonId){
        console.log('llamando', pokemonId, this.pokemon.id)
        this.showPokemon = true
        this.showAnswer = true
        if (pokemonId === this.pokemon.id){
          this.message = `Correcto! es ${ this.pokemon.name}`
        } else {
          this.message = `Ops! era ${ this.pokemon.name}`
        }
      },
      newGame(){
        this.showPokemon = false
        this.showAnswer = false
        this.pokemonArr = []
        this.pokemon = null
        this.mixPokemonArr()
      }
    },
    mounted() {
      this.mixPokemonArr()
    }
}
</script>

<style>

</style>