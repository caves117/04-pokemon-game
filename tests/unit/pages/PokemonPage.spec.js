import {shallowMount, mount} from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage'
import {pokemons} from '../mocks/pokemons.mock'

describe('PokemonPage Component', () => {

    let wrapper
    beforeEach( () => {
        wrapper = shallowMount(PokemonPage)
    })
    
    test('debe de hacer match con el snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('debe de llamar mixPokemonArr al montar', () => {
        const mixPokemonArrSpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArr')
        //se llama a otro wrapper porque se necesita escuchar si se ejecutó el método al montar la aplicación, si se usa el declarado global, para este punto el método ya ha sido llamado
        const wrapper = shallowMount(PokemonPage)
        expect(mixPokemonArrSpy).toHaveBeenCalled()
    })

    test('debe de hacer match con el snapshot cuando se cargan los pokemons', () => {
        //shallowMount - monta de una manera ligera o simulada de los componentes (stub)
        //mount - monto todo los componentes como debe ser
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('Debe de mostrar los componentes de Pokemon Picture y PokemonOptions', () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })
        const picture = wrapper.find('pokemon-picture-stub')
        const options = wrapper.find('pokemon-options-stub')
        //PokemonPicture y PokemonOptions deben de existir
        expect(picture.exists()).toBeTruthy()
        expect(options.exists()).toBeTruthy()
        //PokemonPicture attribute pokemonId === 5
        expect(picture.attributes('pokemonid')).toBe('1')
        //PokemonOptions attribute pokemon toBe true
        expect(options.attributes('pokemons')).toBeTruthy()
    })

    test('pruebas con checkanswer', async() => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })
        await wrapper.vm.checkAnswer(1)
        expect(wrapper.find('h2').exists()).toBeTruthy()
        expect(wrapper.vm.showPokemon).toBeTruthy()
        expect(wrapper.find('h2').text()).toBe(`Correcto! es ${ pokemons[0].name }`)

        await wrapper.vm.checkAnswer(10)
        expect(wrapper.vm.message).toBe(`Ops! era ${ pokemons[0].name}`)
    })

})