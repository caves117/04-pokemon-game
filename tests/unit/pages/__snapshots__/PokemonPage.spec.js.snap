// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component debe de hacer match con el snapshot 1`] = `<h1> Wait</h1>`;

exports[`PokemonPage Component debe de hacer match con el snapshot cuando se cargan los pokemons 1`] = `
<div>
  <h1>¿Quién es este pokemón?</h1>
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
